### PR TITLE
Support Hierarchical Cohorts in hierarchy.Manager

### PR DESCRIPTION
**Pull request by gabesaba**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This will be used by `cache` and `queue` packages as part of HierarchicalCohorts (#79) implementation.

#### Special notes for your reviewer:
Sending this out as a separate PR to keep PR size small.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Review by mimowo — COMMENTED** [review]
LGTM. I would yet like to check the test coverage for the hierarchy package as it is hard for me to assess the level of coverage with the new test cases.

File: pkg/hierarchy/cohort.go [hunk 1]
```diff
@@ -19,31 +19,52 @@ package hierarchy
 import "k8s.io/apimachinery/pkg/util/sets"
 
 type Cohort[CQ, C nodeBase] struct {
-	childCqs sets.Set[CQ]
+	parent       C
```
[the remark this answers is not in the document]
Comment (gabesaba) [reply pkg/hierarchy/cohort.go:22]: There was just one branch missing coverage - fixed with new test "adding cohorts idempotent"

File: pkg/hierarchy/manager.go [hunk 2]
```diff
@@ -94,6 +115,16 @@ func (c *Manager[CQ, C]) unwireClusterQueue(cq CQ) {
 	}
 }
 
+func (c *Manager[CQ, C]) unwireCohort(cohort C) {
```
[the remark this answers is not in the document]
Comment (gabesaba) [reply pkg/hierarchy/manager.go:118]: Done.

File: pkg/hierarchy/manager.go [hunk 3]
```diff
@@ -76,12 +93,16 @@ func (c *Manager[CQ, C]) DeleteCohort(name string) {
 }
 
 // rewireChildren is used when we are changing a Cohort
-// from an implicit to an explicit Cohort, or vice-versa.
+// from an explicit to an implicit Cohort.
 func (c *Manager[CQ, C]) rewireChildren(old, new C) {
```
[the remark this answers is not in the document]
Comment (gabesaba) [reply pkg/hierarchy/manager.go:97]: Done.

File: pkg/hierarchy/manager.go [hunk 4]
```diff
@@ -56,18 +56,35 @@ func (c *Manager[CQ, C]) DeleteClusterQueue(name string) {
 	}
 }
 
-func (c *Manager[CQ, C]) AddCohort(cohort C) {
-	cohort.markExplicit()
-	if oldCohort, ok := c.Cohorts[cohort.GetName()]; ok {
-		c.rewireChildren(oldCohort, cohort)
+func (c *Manager[CQ, C]) AddCohort(cohortName string) {
```
[the remark this answers is not in the document]
Comment (gabesaba) [reply pkg/hierarchy/manager.go:59]: I will send out as a separate PR to keep this diff smaller.

[the remark this answers is not in the document]
Comment (gabesaba) [reply pkg/hierarchy/manager.go:59]: #3002

**Pull request merged**